components/service/glean/src/main/java/mozilla/components/service/glean/storages/TimingDistributionsStorageEngine.kt

```diff
@@ -0,0 +1,243 @@
+     *
+     * @param metricData the metric information for the timing distribution
+     * @param sample the value to accumulate
+     */
+    @Synchronized
+    fun accumulate(
+        metricData: CommonMetricData,
+        sample: Long,
+        timeUnit: TimeUnit
+    ) {
+        val newTD = TimingDistributionData(category = metricData.category, name = metricData.name,
+            timeUnit = timeUnit)
+        newTD.accumulate(sample)
+        super.recordScalar(metricData, newTD, null) { currentValue, newValue ->
+            currentValue?.let {
+                it.accumulate(newValue.sum)
+                it
+            } ?: newValue
+        }
+    }
+
+    /**
+     * Get a snapshot of the stored data as a JSON object.
+     *
+     * @param storeName the name of the desired store
+     * @param clearStore whether or not to clearStore the requested store
+     *
+     * @return the [JSONObject] containing the recorded data.
+     */
+    override fun getSnapshotAsJSON(storeName: String, clearStore: Boolean): Any? {
+        return getSnapshot(storeName, clearStore)?.let { dataMap ->
+            val jsonObj = JSONObject()
+            dataMap.forEach {
+                jsonObj.put(it.key, it.value.toJsonObject())
+            }
+            return jsonObj
+        }
+    }
+}
+
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+data class TimingDistributionData(
+    val category: String,
+    val name: String,
+    val bucketCount: Int = DEFAULT_BUCKET_COUNT,
+    val range: List<Long> = listOf(DEFAULT_RANGE_MIN, DEFAULT_RANGE_MAX),
+    val histogramType: HistogramType = HistogramType.Exponential,
+    val values: MutableMap<String, Long> = mutableMapOf(),
+    var underflow: Long = 0,
+    var overflow: Long = 0,
+    var sum: Long = 0,
+    val timeUnit: TimeUnit = TimeUnit.Millisecond
+) {
+    companion object {
+        // These represent the indexes of the min and max of the timing distribution range
+        const val MIN = 0
+        const val MAX = 1
+
+        // The following are defaults for a simple timing distribution
+        const val DEFAULT_BUCKET_COUNT = 50
+        const val DEFAULT_RANGE_MIN = 0L
+        const val DEFAULT_RANGE_MAX = 30000L
+
+        /**
+         * Factory function that takes stringified JSON and converts it back into a
+         * [TimingDistributionData].  This tries to read all values and attempts to
+         * use a default where no value exists.
+         *
+         * @param json Stringified JSON value representing a [TimingDistributionData] object
+         * @return A [TimingDistributionData] or null if unable to rebuild from the string.
+         */
+        fun fromJsonString(json: String): TimingDistributionData? {
+            val jsonObject: JSONObject
+            try {
+                jsonObject = JSONObject(json)
+            } catch (e: org.json.JSONException) {
+                return null
+            }
+
+            return TimingDistributionData(
```

Comment:
Yeah, I think I agree.  Normally I'd say "be liberal in what you accept", but since this is data our own library has written to disk, I think we should be more skeptical and throw data away that looks fishy, and maybe report an error metric.  (We might need a new error metric type for this to indicate "corrupted data" or something).